x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/tokenizers/BasicTokenizer.java

```diff
@@ -0,0 +1,273 @@
+     * combining characters are arranged in a specific order"
+     * from https://en.wikipedia.org/wiki/Unicode_equivalence#Normal_forms
+     *
+     * And remove non-spacing marks https://www.compart.com/en/unicode/category/Mn
+     *
+     * @param word Word to strip
+     * @return {@code word} normalized and stripped.
+     */
+    static String stripAccents(String word) {
+        String normalizedString = Normalizer.normalize(word, Normalizer.Form.NFD);
+
+        int [] codePoints = normalizedString.codePoints()
+            .filter(codePoint -> Character.getType(codePoint) != Character.NON_SPACING_MARK)
+            .toArray();
+
+        return new String(codePoints, 0, codePoints.length);
+    }
+
+    static List<String> splitOnPunctuation(String word) {
+        List<String> split = new ArrayList<>();
+        int [] codePoints = word.codePoints().toArray();
+
+        int lastSplit = 0;
+        for (int i=0; i<codePoints.length; i++) {
+            if (isPunctuation(codePoints[i])) {
+                int charCount = i - lastSplit;
+                if (charCount > 0) {
+                    // add a new string for what has gone before
+                    split.add(new String(codePoints, lastSplit, i - lastSplit));
+                }
+                split.add(new String(codePoints, i, 1));
+                lastSplit = i+1;
+            }
+        }
+
+        if (lastSplit < codePoints.length) {
+            split.add(new String(codePoints, lastSplit, codePoints.length - lastSplit));
+        }
+
+        return split;
+    }
+
+    /**
+     * Surrounds any CJK character with whitespace
+     * @param text To tokenize
+     * @return tokenized text
+     */
+    static String tokenizeCjkChars(String text) {
+        StringBuilder sb = new StringBuilder(text.length());
+        boolean cjkCharFound = false;
+
+        int[] codePoints = text.codePoints().toArray();
+        for (int cp : codePoints) {
+            if (isCjkChar(cp)) {
+                sb.append(' ');
+                sb.appendCodePoint(cp);
+                sb.append(' ');
+                cjkCharFound = true;
+            } else {
+                sb.appendCodePoint(cp);
+            }
+        }
+
+        // no change
+        if (cjkCharFound == false) {
+            return text;
+        }
+
+        return sb.toString();
+    }
+
+    /**
+     * Remove control chars and normalize white space to ' '
+     * @param text Text to clean
+     * @return Cleaned text
+     */
+    static String cleanText(String text) {
+        int [] codePoints = text.codePoints()
+            .filter(codePoint -> (codePoint == 0x00 || codePoint == 0xFFFD || isControlChar(codePoint)) == false)
+            .map(codePoint -> isWhiteSpace(codePoint) ? ' ' : codePoint)
+            .toArray();
+
+        return new String(codePoints, 0, codePoints.length);
+    }
+
+    static boolean isCjkChar(int codePoint) {
```

Comment:
Could you use `Character.UnicodeBlock.of(codePoint);` in conjunction with `            Character.UnicodeBlock.CJK_COMPATIBILITY_IDEOGRAPHS` ?

Reply:
Character.UnicodeBlock.CJK_COMPATIBILITY_IDEOGRAPHS corresponds to the range `(codePoint >= 0xF900 && codePoint <= 0xFAFF)` 

This may not be optimal but it captures the idiosyncrasies of the original implementation